networkx/algorithms/isomorphism/tree_isomorphism.py

```diff
@@ -1,279 +1,686 @@
 
-    Note that two trees may have more than one isomorphism, and this
-    routine just returns one valid mapping.
+    For each level i in T1, check that all the structures present on the i-th
+    level are also present in i-th level of T2. If all the levels are the same
+    for both trees, return true.
 
     Parameters
     ----------
-    t1 : undirected NetworkX graph
-        One of the trees being compared
+    T1 : NetworkX Graph
+        The rooted tree T1.
 
-    t2 : undirected NetworkX graph
-        The other tree being compared
+    root_T1 : node
+        The root of the rooted tree T1.
+
+    T2 : NetworkX Graph
+        The rooted tree T2.
+
+    root_T2 : node
+        The root of the rooted tree T2.
 
     Returns
     -------
-    isomorphism : list
-        A list of pairs in which the left element is a node in `t1`
-        and the right element is a node in `t2`.  The pairs are in
-        arbitrary order.  If the nodes in one tree is mapped to the names in
-        the other, then trees will be identical. Note that an isomorphism
-        will not necessarily be unique.
+    are_iso : bool
+        Returns True if the given rooted trees are isomorphic, False otherwise.
 
-        If `t1` and `t2` are not isomorphic, then it returns the empty list.
+    isomorphism : dict of node of T1: node of T2
+        The isomorphism mapping f such that (u, v) is in E_T1 if and only if
+        (f(u), f(v)) is in E_T2. This dictionary is empty when there is no
+        isomorphism between the trees.
+
+    Note
+    ----
+    This algorithm runs in O(n) time and space.
+
+    """
+    # Get the intial maps and height for each tree.
+    (
+        levels_T1,
+        values_T1,
+        children_T1,
+        parenthood_T1,
+        height_T1,
+    ) = get_initial_maps_and_height(T1, root_T1)
+
+    (
+        levels_T2,
+        values_T2,
+        children_T2,
+        parenthood_T2,
+        height_T2,
+    ) = get_initial_maps_and_height(T2, root_T2)
+
+    # If the trees differ in height, return false.
+    if height_T1 != height_T2:
+        return False, {}
+
+    # If both trees have height 0, then there are only conformed by its roots,
+    # they're isomorphic.
+    if height_T1 == 0:
+        return True, {root_T1: root_T2}
+
+    # For every level i, check that both trees have the same amount of
+    # vertices on the i-th level. If they differ for some level, return
+    # false.
+    for i in range(height_T1):
+        if len(levels_T1[i]) != len(levels_T2[i]):
+            return False, {}
+
+    # Start at level 1 as all the vertices at level 0 are leaves. Check that
+    # all the structures present on the i-th level are also present in i-th
+    # level of T2.
+    for current_level in range(1, height_T1 + 1):
+        # Assign the structures for the vertices of the current level for T1 and
+        # T2.
+        struct_T1 = {
+            v: Counter([values_T1[u] for u in children_T1[v]])
```

Comment:
You don't need a list here since `Counter` accepts iterators as arguments (I think)